apis/mutations/unversioned/externaldata.go

```diff
@@ -0,0 +1,75 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package unversioned
+
+import (
+	"errors"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/externaldata"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/types"
+)
+
+var (
+	ErrExternalDataFeatureFlag = errors.New("external data feature flag is not enabled")
+	ErrExternalDataNoDefault   = errors.New("external data must set `default` when `failurePolicy` is `UseDefault`")
+)
+
+// ExternalData describes the external data source to use for the mutation.
+type ExternalData struct {
+	// Provider is the name of the external data provider.
+	// +kubebuilder:validation:Required
+	Provider string `json:"provider,omitempty"`
+
+	// DataSource specifies where to extract the data that will be sent
+	// to the external data provider as parameters.
+	// +kubebuilder:default="ValueAtLocation"
+	DataSource types.ExternalDataSource `json:"dataSource,omitempty"`
+
+	// FailurePolicy specifies the policy to apply when the external data
+	// provider returns an error.
+	// +kubebuilder:default="Fail"
+	FailurePolicy types.ExternalDataFailurePolicy `json:"failurePolicy,omitempty"`
+
+	// Default specifies the default value to use when the external data
+	// provider returns an error and the failure policy is set to "UseDefault".
+	// +kubebuilder:validation:Schemaless
+	// +kubebuilder:validation:XPreserveUnknownFields
+	Default *types.Anything `json:"default,omitempty"`
+}
+
+// GetPlaceholder returns the placeholder for the given external data provider.
+func (e *ExternalData) GetPlaceholder(username string, mutatorID types.ID) (map[string]interface{}, error) {
```

Comment:
It looks like this function isn't called, is it still necessary?

Reply:
It's called in AssignField's `GetValue()`.

Reply:
Ah, could it also receive `value` as an argument and handle the replacing/merging logic if `value` happens to be a Placeholder{} then?